sample(12): Example of how to use @ResolveField

gabeodess opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
 - [ ] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [X] Other... Please describe: Updates sample code

## What is the current behavior?
There is no demo of `@ResolveField` in the `graphql-schema-first` sample

Issue Number: N/A


## What is the new behavior?
Adds a demo of `@ResolveField`

## Does this PR introduce a breaking change?
 - [ ] Yes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

I love that you've added a GQL schema-first sample!  Huge help to understanding how we can jump right in to NestJS with a little structure.  One challenge my team has  been facing is figuring out the right way to translate information from the data-layer (TypeORM) to the transport-layer (the resolver).  

In the sample you've provided, it looks like the data-layer and the transport-layer are 1-to-1 (the GQL types perfectly match the data-set) so there's no need to solve this problem there.  I would love to see how you would implement an association.  The main structure I would love to see solved is where a foreign-key lives in the data-layer but does not exist in the transport-layer.

This PR gets us most of the way, but there is a step missing to satisfy the typing, and that's what I'm hoping to get some guidance on.  I'm wondering if we should create some kind of DTO and have `CatResolvers.findOneById` return the DTO instead of the GQL type, or if there's a better way to handle this 🤔 

The example here adds an associated `Owner` with a FK on the cat data to link the two.  We would like to add `Cat.owner` without adding `ownerId` to the GQL type.

Limkiathow0626 commented:
patch或.diff

limkiathow2910 commented on sample/12-graphql-schema-first/src/cats/cats.service.ts:
```diff
@@ -1,9 +1,10 @@
 import { Injectable } from '@nestjs/common';
-import { Cat } from '../graphql.schema';
```
O

kamilmysliwiec commented:
Thanks for your contribution @gabeodess! To stay in line with the best practices, can we create a dedicated service (and module) for the "Owner" entity?

gabeodess commented:
Thanks for taking a look @kamilmysliwiec!  I've added the owner module and service.  The big hurtle for me here is how we should go about handling the foreign key.   You can see I've added the FK to the `cats` data-set, however, this is a TS violation since the data set is typed as the GQL type `Cat` which has no FK defined on it.  How would you go about handling this?

TkachenkoDmitry commented:
@gabeodess Thanks for the example! Do you have any plans in creating an example with [dataloader](https://github.com/graphql/dataloader)? 1+N is becoming a problem really quick, and all the solutions for nest are not showing DI or a bit hacky and 'boilerplaty'.

Sorry, for mentioning you directly. Maybe that's the question for @kamilmysliwiec and out of the scope of this PR

gabeodessru commented:
> Do you have any plans in creating an example with dataloader?

That would be a great addition to this sample app, but I think it would be in a separate PR. Personally I haven’t tackled the data loader solution yet so I’m not sure what it would look like.

kamilmysliwiec commented on integration/graphql-schema-first/src/owners/owners.module.ts:
```diff
@@ -0,0 +1,14 @@
+  providers: [OwnersService],
+})
+export class OwnersModule {
+  static enableRequestScope(): DynamicModule {
+    return {
+      module: OwnersModule,
+      providers: [OwnersService],
+    };
+  }
```
Can we remove this method?

gabeodessru replied:
Done!
